### PR TITLE
basic naive search

### PR DESCRIPTION
**Pull request by timgoeller**
Very basic search that iterates messages that fit the opts (the opts that go into `getMessages()`) and returns a promise that will resolve to a list of matches. These matches are objects with a message property and a list of indexes where the search string was found. Messages that don't contain the search string won't be returned.

### CODE REVIEW

File: src/client.js [hunk 1]
```diff
@@ -416,6 +416,58 @@ class Client {
...
+   * @param {number} [opts.amount] amount of messages to be search through
+   * @param {string} [opts.channel] channel to get messages from. defaults to currently focused channel
+   * @param {Cabal} [cabal=this.currentCabal]
+   */
+  searchMessages(searchString, opts, cabal = this.currentCabal) {
+    return new Promise((resolve, reject) =>
+      {
+        if(!searchString && searchString != "") {
+          reject('search string must be set')
+        }
```
Comment (cblgh) [inline src/client.js:434]:
```suggestion
        if (!searchString && searchString !== "") {
          return reject('search string must be set')
        }
```

File: src/client.js [hunk 2]
```diff
@@ -416,6 +416,58 @@ class Client {
...
+        this.getMessages(opts, null, cabal).then((messages) =>
+          {
+            messages.forEach(message => {
+              let messageContent = message.value.content
+              if(messageContent) {
+                let textBuffer = Buffer.from(messageContent.text)
+
+                let matchedIndexes = []
+
+                charIteration:
```
Comment (cblgh) [inline src/client.js:449]: woah never seen a named/labeled `for`-loop before

File: src/client.js [hunk 3]
```diff
@@ -416,6 +416,58 @@ class Client {
...
+   * @param {Cabal} [cabal=this.currentCabal]
+   */
+  searchMessages(searchString, opts, cabal = this.currentCabal) {
+    return new Promise((resolve, reject) =>
+      {
+        if(!searchString && searchString != "") {
+          reject('search string must be set')
+        }
+
+        let searchBuffer = Buffer.from(searchString)
```
Comment (cblgh) [inline src/client.js:436]:
```suggestion
      const searchBuffer = Buffer.from(searchString)
```

File: src/client.js [hunk 4]
```diff
@@ -416,6 +416,58 @@ class Client {
...
+
+        let matches = []
```
Comment (cblgh) [inline src/client.js:438]:
```suggestion
         const matches = []
```

File: src/client.js (hunk 2 above)
Comment (cblgh) [inline src/client.js:449]:
```suggestion
                 /* use a labeled for-loop to cleanly continue top-level iteration */
                 charIteration:
```

File: src/client.js [hunk 5]
```diff
@@ -416,6 +416,58 @@ class Client {
...
+                      for (let searchIndex = 0; searchIndex < searchBuffer.length; searchIndex++) {
+                        if(!(textBuffer[charIndex + searchIndex] == searchBuffer[searchIndex])) 
+                          continue charIteration             
+                      }
+                      matchedIndexes.push(charIndex)
+                    }
+                  }
+                
+                  if(matchedIndexes.length > 0) {
+                    matches.push({message: message, matchedIndexes: matchedIndexes})
```
Comment (cblgh) [inline src/client.js:461]:
```suggestion
                    matches.push({ message, matchedIndexes })
```

File: src/client.js (hunk 1 above)
Comment (timgoeller) [reply src/client.js:434]: Does the `return`  here have a function, or did you add it to clarify that the method exits there?

File: src/client.js (hunk 5 above)
Comment (cblgh) [reply src/client.js:461]: after reading more closely, maybe you can comment a bit inline on the code what the intent is of the `matchedIndexes` array?

is it to be able to highlight the string at parts of the matched message?

File: src/client.js (hunk 1 above)
Comment (cblgh) [reply src/client.js:434]: iirc i think execution will continue without it? *goes to check if i'm actually mistaken lol*

Comment (cblgh) [reply src/client.js:434]: https://gist.github.com/cblgh/6b98da32d30550e181305fe6e6bd61f2 looks like i'm not mistaken, try this gist

File: src/client.js (hunk 5 above)
Comment (timgoeller) [reply src/client.js:461]: Yes, and to give the client more flexibility to decide what to do with the results. It contains all positions in the string where the search string was found. Could also be useful if you only want three characters surrounding the result(s), or things like that for example.

I can add a comment to clarify that!

Comment (cblgh) [reply src/client.js:461]: fantastic, a clarifying comment would be 🍏

File: src/client.js (hunk 1 above)
Comment (timgoeller) [reply src/client.js:434]: Huh, I didn't know that :O. Not sure why I expected otherwise since `reject` is just a function call

**Comment by timgoeller** [comment]
Didn't know about `standard`  before, thanks for the hint!

**Comment by cblgh** [comment]
@timgoeller looks great to me! thank you so much! this closes #5 and helps unblock https://github.com/cabal-club/cabal-desktop/issues/143

can you generate the docs as well? just run [`npm run doc`](https://github.com/cabal-club/cabal-client/blob/master/package.json#L15) and the docstrings you added will added to https://github.com/cabal-club/cabal-client/blob/master/api.md

**Pull request merged**